### DISCUSSION TITLE
Change not reflected using `setVal()`

### DISCUSSION THREAD

**Discussion by etkaar** [HomeSpan/HomeSpan#868] [Q&A]
For a bistable relay I have a status pin in order to find out what state the relays has on restart. I check this value in loop() and update it by using `setVal(1, true)` or `setVal(0, true)`, but the change it not reflected in HomeKit. First I thought, it is due to the WiFi connection, as I can see that it will fire _before_ the WiFi connection is established (is that normal?), which can be seen here:

<img width="503" alt="image" src="https://github.com/HomeSpan/HomeSpan/assets/40885610/48673533-8e5c-4ff9-94e7-7d9bbb4ac68c">

But even a WiFi.status() == WL_CONNECTED did not help. It then fires _after_ the WiFi connection has been established,

<img width="469" alt="image" src="https://github.com/HomeSpan/HomeSpan/assets/40885610/d27047e1-f6ff-4d09-983f-b3af83442abd">

but still the change is not reflected in HomeKit. I can see, however, that there is no time in the log. Does it have something to do with NTP? Or any other idea why it is not reflected?

```
struct DEV_CALL_SWITCH : Service::Switch {

  uint8_t index;
  uint8_t pin_number_enable;
  uint8_t pin_number_disable;
  uint8_t pin_number_status;

  SpanCharacteristic *element_call;

  DEV_CALL_SWITCH(uint8_t index, uint8_t pin_number_enable, uint8_t pin_number_disable, uint8_t pin_number_status) : Service::Switch() {
    this->index = index;
    this->pin_number_enable = pin_number_enable;
    this->pin_number_disable = pin_number_disable;
    this->pin_number_status = pin_number_status;

    element_call = new Characteristic::On();

    pinMode(this->pin_number_enable, OUTPUT);
    pinMode(this->pin_number_disable, OUTPUT);
    pinMode(this->pin_number_status, INPUT);

    WEBLOG("Call %d: Set enable pin to OUTPUT for GPIO%d.", index, pin_number_enable);
    WEBLOG("Call %d: Set disable pin to OUTPUT for GPIO%d.", index, pin_number_disable);
    WEBLOG("Call %d: Set status pin to INPUT for GPIO%d.", index, pin_number_status);
  }

  bool update() override {
    if (element_call->getNewVal() == 1) {
      enable();
      WEBLOG("Call %d: Enabled.", index);
    } else {
      disable();
      WEBLOG("Call %d: Disabled.", index);
    }

    return true;
  }

  void loop() override {
    if (WiFi.status() == WL_CONNECTED) {
      if (is_relay_set() && element_call->getVal() == 0) {
        element_call->setVal(1, true);
        WEBLOG("Call %d: Synchronize with relay state (ON).", index);
      } else if (!is_relay_set() && element_call->getVal() == 1) {
        element_call->setVal(0, true);
        WEBLOG("Call %d: Synchronize with relay state (OFF).", index);
      }
    }
  }

  bool is_relay_set() {
    if (digitalRead(this->pin_number_status) == HIGH) {
      return true;
    }

    return false;
  }

  void enable() {
    digitalWrite(this->pin_number_enable, HIGH);
    delay(50);
    digitalWrite(this->pin_number_enable, LOW);
  }

  void disable() {
    digitalWrite(this->pin_number_disable, HIGH);
    delay(50);
    digitalWrite(this->pin_number_disable, LOW);
  }

};

```

**Comment by etkaar** [comment]
Does anyone have an idea?

**Comment by HomeSpan** [comment]
Changes can be made before or after WiFi is established - should not matter because if you make a change before WiFi is established, HomeSpan will read the latest value of the Characteristic once the Home App connects for the first time.

In general, when initializing a status variable at start-up I usually put the logic directly into the constructor.  Something like:

```C++
pinMode(this->pin_number_status, INPUT);
Serial.printf("Pin status is: %d\n", digitalRead(this->pin_number_status));
element_call = new Characteristic::On( digitalRead(this->pin_number_status ));
```

**Reply by etkaar** [reply]
Hmm, but why is the change not reflected? As you can see in `loop()` it constantly checks if the relay state has changed, so it makes sure it is in synch. But the call `element_call->setVal([bool], true);` has no effect for me.

**Comment by HomeSpan** [comment]
Are you using a Home Hub (either an Apple TV or Home Pod)?

**Reply by etkaar** [reply]
Yes, an Apple TV.

**Comment by HomeSpan** [comment]
Can you test with a clean install of [Example 12](https://github.com/HomeSpan/HomeSpan/tree/master/examples/12-ServiceLoops).  This simulates a temperature sensor and an air-quality sensor.  It should update the values every 5 seconds or so within the Home App.  I just check and confirmed it works.
